### PR TITLE
From-string object transformer

### PR DESCRIPTION
**Pull request by everzet**
## TLDR;

This PR introduces automatic ValueObject transformer into Behat. It is first half of #863. It has following attributes:

1. Is automatically executed for all your class type-hints in your step definitions
2. Attempts to automatically transform such arguments calling typehint's `fromString()` factory method
3. Has the lowest precedence of all transformers. Meaning that if you provide any other kind of transformation that matches said argument, from-string object transformer wouldn't get executed

### Feature

Given simple feature:

```gherkin
Feature:
  Scenario:
    Given I am "everzet"
```

### Stepdef

And a step definition, using type-hint:

```php
class User {
    public function __construct($name) { $this->name = $name; }
}

class FeatureContext implements Context
{
    /** @Given I am :user */
    public function iAm(User $user) {
        // TODO: do something with $user
    }
}
```

### Test

When you attempt to execute this, you'll get exception like this:

```bash
...> Method fromString does not exist
```

There are now two possible routes out of this situation...

#### Route 1: Write custom transformation

As before, you can write your own custom transformation and it will work as it always did:

```php
class User {
    public function __construct($name) { $this->name = $name; }
}

class FeatureContext implements Context
{
    /** @Transform :user */
    public function transformUser($user) {
        return new User($user);
    }

    /** @Given I am :user */
    public function iAm(User $user) {
        // TODO: do something with $user
    }
}
```

#### Route 2: Add `fromString`

Or, you can add `fromString` factory method to your `User` class and that will also work exactly as you might expect:

```php
class User {
    private function __construct($name) { $this->name = $name; }
    static public function fromString($name) { return new static($name); }
}

class FeatureContext implements Context
{
    /** @Given I am :user */
    public function iAm(User $user) {
        // TODO: do something with $user
    }
}
```

## Implications

There are no known BC breaks or implications on existing codebases. Please, highlight if you are aware of any.

### CODE REVIEW

**Comment by ciaranmcnulty** [comment]
While I would use this as an extension, I don't think this should be core.

Thus far Behat has not introduced any conventions to how users design or write code, and I don't think it's appropriate for the core app to do so.

If you want to go ahead, the error message needs to be far more explicit about why the hell it things `fromString` should exist or why it's looking for it.

**Comment by everzet** [comment]
> Thus far Behat has not introduced any conventions to how users design or write code, and I don't think it's appropriate for the core app to do so.
Why?

File: src/Behat/Behat/Transformation/Call/Filter/DefinitionArgumentsTransformer.php [hunk 1]
```diff
@@ -100,7 +100,11 @@ private function transformArgument(DefinitionCall $definitionCall, $index, $valu
                 continue;
             }
 
-            return $transformer->transformArgument($definitionCall, $index, $value);
+            $transformedValue = $transformer->transformArgument($definitionCall, $index, $value);
+
+            if ($value !== $transformedValue) {
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/Transformation/Call/Filter/DefinitionArgumentsTransformer.php]: This condition is not needed. If they're equal you can still return `$transformedValue` (as `$value` is returned below);

Additionally the `continue` above doesn't add much to comprehension, you could flip that `if` and nest the `return` inside it.

File: src/Behat/Behat/Transformation/Exception/FactoryMethodNotFound.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * @var string
+     */
+    private $method;
+
+    /**
+     * Initializes exception.
+     *
+     * @param string $message
+     * @param string   $class
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/Transformation/Exception/FactoryMethodNotFound.php]: Missing a  docblock for`$method`

File: src/Behat/Behat/Transformation/Exception/FactoryMethodNotFound.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        return $this->class;
+    }
+
+    /**
+     * Return factory method name.
+     *
+     * @return string
+     */
+    public function getFactoryMethod()
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/Transformation/Exception/FactoryMethodNotFound.php]: If it's a factoryMethod then the constructor param should be named the same

File: src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+use ReflectionParameter;
+
+/**
+ * Transforms typehinted parameters using their `fromString` factory method (if exists).
+ *
+ * @author Konstantin Kudryashov <ever.zet@gmail.com>
+ */
+final class FromStringObjectTransformer implements ArgumentTransformer
+{
+    const FACTORY_METHOD = 'fromString';
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php]: Could this be passed into the constructor instead? It could come from configuration but have a default?

File: src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function transformArgument(DefinitionCall $definitionCall, $argumentIndex, $argumentValue)
+    {
+        $class = $this->getParameterClassByIndex($definitionCall, $argumentIndex);
+
+        if (!$class->hasMethod(self::FACTORY_METHOD)) {
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php]: Probably also need to check the method is compatible i.e. Unary, can take a string as it's arguments (compatible typehints)

**Comment by ciaranmcnulty** [comment]
New exception message is a big improvement

File: src/Behat/Behat/Transformation/Call/Filter/DefinitionArgumentsTransformer.php (hunk 1 above)
Comment (everzet) [inline src/Behat/Behat/Transformation/Call/Filter/DefinitionArgumentsTransformer.php]: I don't want successful transformations to stack.

File: src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php (hunk 5 above)
Comment (everzet) [inline src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php]: That's why I don't want it to be configurable and anything but `fromString`. If you want to configure it - you should use custom transformation.

File: src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php (hunk 4 above)
Comment (everzet) [inline src/Behat/Behat/Transformation/Transformer/FromStringObjectTransformer.php]: If you want to configure it - you should use custom transformation.

File: src/Behat/Behat/Transformation/Exception/FactoryMethodNotFound.php (hunk 3 above)
Comment (everzet) [inline src/Behat/Behat/Transformation/Exception/FactoryMethodNotFound.php]: Ta. Fixed.

**Comment by jakzal** [comment]
It doesn't feel right to introduce the `fromString` as a convention. It will encourage adding this kind of method to all value objects (since transformers will become so easy). In general I share @ciaranmcnulty's concerns expressed in #863. 

I really like option 3 from https://github.com/Behat/Behat/issues/863#issuecomment-235400924. Any reason why we couldn't support a variation of that in PHP 5.x?

**Comment by pamil** [comment]
Totally agree with @jakzal and @ciaranmcnulty, especially that with the second or third option from https://github.com/Behat/Behat/issues/863#issuecomment-235400924 implemented, it will be a no-brainer (and ~6 lines of code) to implement own type-based transformations.

Support for PHP 5.x will be very appreciated, we use a lot of transformers that can be type-based in [Sylius](https://github.com/Sylius/Sylius/tree/master/src/Sylius/Behat/Context/Transform) and can't bump up dependencies to `^7.0`.

**Comment by stof** [comment]
> When you attempt to execute this, you'll get exception like this:
> 
> ...> Method fromString does not exist
> 
> There are now two possible routes out of this situation...

I think that this message is not clear at all. I think it would be better if the transformation was checking the existence of ``fromString`` to decide whether it supports the argument or no, which would result in the argument not being transformed if it cannot.
Thus, it would remove the requirement to keep this transformer in last position forever (the current implementation requires it because if partial support check)

**Comment by everzet** [comment]
Ok, it seems we're overwhelmingly against this feature in its current form. I then see two options for this PR:

1. Close it like it never happened
2. Do what @stof suggested. Remove the exception throwing by silently checking for `fromString` existence before transformation. If method exists - automatically use it for transformation, if not - silently skip the transformation

Thoughts?

**Comment by everzet** [comment]
@jakzal, @pamil I'd keep discussion about wider object transformations improvement outside of the scope of this PR and keep it in the #863.
